### PR TITLE
Adds a test that should break

### PR DESCRIPTION
**Pull request by ianoc**
This should break in the ordered serialization

I've hit exceptions from it like OOM's, missing serializers for BoxedN and so on.

### CODE REVIEW

**Comment by MansurAshraf** [comment]
+1

File: scalding-hadoop-test/src/test/scala/com/twitter/scalding/ordered_serialization/OrderedSerializationTest.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  def sample[T: Arbitrary]: T = Arbitrary.arbitrary[T].sample.get
+  val data = sample[List[NestedCaseClass]].take(1000)
+}
+
+case class NestedCaseClass(day: RichDate, key: (String, String))
+
+class OrderedSerializationTest extends FunSuite with HadoopPlatformTest {
+  import OrderedSerializationTest._
+  test("A test job with a fork and join, had previously not had boxed serializations on all branches") {
```
Comment (johnynek) [inline scalding-hadoop-test/src/test/scala/com/twitter/scalding/ordered_serialization/OrderedSerializationTest.scala]: this is only testing for no exception, right? Can we add a comment so we don't lose track of that later (or add also a logic test of what the results should be)?

File: scalding-core/src/main/scala/com/twitter/scalding/FlowState.scala [hunk 2]
```diff
@@ -21,9 +21,12 @@ import java.util.WeakHashMap
 /**
  * Immutable state that we attach to the Flow using the FlowStateMap
  */
-case class FlowState(sourceMap: Map[String, Source] = Map.empty) {
+case class FlowState(sourceMap: Map[String, Source] = Map.empty, flowBoxes: Set[(String, String)] = Set()) {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/FlowState.scala]: what about changing flowBoxes to `flowConfigUpdates` or something. It seems like it is not really to do with the serialization and could be used for any global flow config key/value updates.

File: scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala [hunk 3]
```diff
@@ -660,6 +661,45 @@ class RichPipe(val pipe: Pipe) extends java.io.Serializable with JoinAlgorithms
       .flatten
       .toSet
 
+  /**
+   * This finds all the boxed serializations stored in the flow state map for this
+   * flowdef. We then find all the pipes back in the DAG from this pipe and apply
+   * those serializations.
+   */
+  private[scalding] def applyBoxedSerializations(flowDef: FlowDef): Pipe = {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala]: can we change this to `applyFlowConfigProperties` or something?

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala [hunk 4]
```diff
@@ -93,46 +93,24 @@ object Grouped {
    * If we are using OrderedComparable, we need to box the key
    * to prevent other serializers from handling the key
    */
-  def maybeBox[K, V](ord: Ordering[K])(op: (TupleSetter[(K, V)], Fields) => Pipe): Pipe = ord match {
+  def maybeBox[K, V](ord: Ordering[K], flowDef: FlowDef)(op: (TupleSetter[(K, V)], Fields) => Pipe): Pipe = ord match {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Grouped.scala]: should this be `private[scalding]`? Seems super low level to me.

**Comment by johnynek** [comment]
Aside from some minor naming comments, this looks great. Glad you got it sorted.

**Pull request merged**